### DISCUSSION TITLE
Integrating MSG91 with Supabase

### DISCUSSION THREAD

**Discussion by Saipreetham0** [supabase/supabase#30197] [Feature Requests] [answered]
Hey everyone! 👋

I’ve been exploring the possibility of integrating MSG91 (for SMS services) with Supabase to manage real-time messaging, user verifications, or notifications within our app. Since Supabase doesn’t have built-in support for MSG91, I’m considering possible methods to make this work.

Has anyone tried connecting MSG91 with Supabase before? What would be the best way to go about it—maybe using Supabase's serverless functions or APIs?

Any suggestions, guides, or code snippets would be really helpful! 😊

Thanks in advance!

**Comment by alrahimkhanrj** [comment, marked as answer]
I'm also trying to do the same, Supabase has something called SMS Hook, hope this helps you. Let me know how to integrate MSG91 with Supabase for OTP account authentication
https://supabase.com/docs/guides/auth/auth-hooks/send-sms-hook?queryGroups=language&language=http

**Comment by mnvsk97** [comment]
Hello @alrahimkhanrj @Saipreetham0 
Were you guys able to figure out how to integrate msg91? I came across this issue today and this is the closest I've come to understanding how to use both of them together. Happy collaborate if it helps.

**Reply by kaviyarasumaran** [reply]
hi @mnvsk97 @alrahimkhanrj, i also try to connect the msg91 in sms hook, but i face the issue of session management if any one know about how to authenticate using smg91 in supabase 


In this edge is send sms and verify the sms bit session is not activated





import { createClient } from "https://esm.sh/@supabase/supabase-js@2";

// CORS headers
const CORS_HEADERS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "POST, OPTIONS",
  "Access-Control-Allow-Headers": "Content-Type, Authorization",
};

// Initialize Supabase client
const supabaseUrl = Deno.env.get("SUPABASE_URL");
const supabaseKey = Deno.env.get("SUPABASE_SERVICE_ROLE_KEY");
const supabase = createClient(supabaseUrl, supabaseKey);

const sendTextMessage = async (
  messageBody: string,
  toNumber: string,
  otp: string
): Promise<any> => {
  const authKey = Deno.env.get("MSG91_AUTH_KEY");
  const templateId = Deno.env.get("TEMPLATE_ID");

  if (!authKey || !templateId) {
    throw new Error("Missing MSG91_AUTH_KEY or TEMPLATE_ID environment variables");
  }

  const url = "https://control.msg91.com/api/v5/flow";
  
  const payload = {
    template_id: templateId,
    recipients: [{
      mobiles: toNumber.startsWith("91") ? toNumber : `91${toNumber}`,
      var1: otp,
      var2: otp
    }]
  };

  console.log("Sending SMS with payload:", JSON.stringify(payload));

  const response = await fetch(url, {
    method: "POST",
    headers: {
      "authkey": authKey,
      "Content-Type": "application/json",
      "accept": "application/json"
    },
    body: JSON.stringify(payload)
  });

  const responseData = await response.json();
  console.log("MSG91 API response:", responseData);

  if (!response.ok) {
    throw new Error(`MSG91 API error: ${responseData.message || response.statusText}`);
  }

  return responseData;
};

// Simple JWT-like token generator (replace with proper JWT in production)
const generateToken = (phone: string): string => {
  const header = btoa(JSON.stringify({ alg: "HS256", typ: "JWT" }));
  const payload = btoa(JSON.stringify({ phone, iat: Date.now() / 1000 }));
  const signature = btoa("simple-signature"); // Replace with proper signing in production
  return `${header}.${payload}.${signature}`;
};

Deno.serve(async (req) => {
  if (req.method === "OPTIONS") {
    return new Response(null, {
      status: 204,
      headers: { ...CORS_HEADERS },
    });
  }

  if (req.method !== "POST") {
    return new Response(
      JSON.stringify({ error: { http_code: 405, message: "Method Not Allowed" } }),
      {
        status: 405,
        headers: { "Content-Type": "application/json", ...CORS_HEADERS },
      }
    );
  }

  try {
    const payload = await req.json();
    console.log("Request payload:", payload);

    const { phone, action, otp } = payload;

    if (!phone) {
      throw new Error("Missing phone number in payload");
    }

    const normalizedPhone = phone.startsWith("91") ? phone : `91${phone}`;

    if (action === "send") {
      // Generate a 4-digit OTP
      const generatedOtp = Math.floor(1000 + Math.random() * 9000).toString();
      const messageBody = `Dear Customer, Your Verification Code For Company name Is ${generatedOtp}. Please do Not share this Code with anyone.`;

      const responseData = await sendTextMessage(messageBody, normalizedPhone, generatedOtp);

      if (responseData.type === "success") {
        // Store OTP in Supabase under ayul schema
        const expiresAt = new Date(Date.now() + 5 * 60 * 1000).toISOString(); // 5 minutes expiry
        const { error } = await supabase
          .schema("ayul")
          .from("otp_verifications")
          .upsert(
            { phone: normalizedPhone, otp: generatedOtp, expires_at: expiresAt },
            { onConflict: "phone" } // Upsert based on phone
          );

        if (error) {
          throw new Error(`Failed to store OTP: ${error.message}`);
        }

        console.log("Stored OTP for", normalizedPhone, ":", generatedOtp);

        return new Response(
          JSON.stringify({ success: true, message: "SMS sent successfully" }),
          {
            status: 200,
            headers: { "Content-Type": "application/json", ...CORS_HEADERS },
          }
        );
      } else {
        throw new Error(`MSG91 response error: ${responseData.message}`);
      }
    } else if (action === "verify") {
      if (!otp) {
        throw new Error("Missing OTP in payload");
      }

      // Fetch OTP from Supabase
      const { data, error } = await supabase
        .schema("ayul")
        .from("otp_verifications")
        .select("otp, expires_at")
        .eq("phone", normalizedPhone)
        .single();

      if (error || !data) {
        throw new Error("No OTP found for this phone number");
      }

      const { otp: storedOtp, expires_at } = data;
      const isExpired = new Date(expires_at) < new Date();

      if (isExpired) {
        await supabase
          .schema("ayul")
          .from("otp_verifications")
          .delete()
          .eq("phone", normalizedPhone);
        throw new Error("OTP has expired");
      }

      if (storedOtp !== otp) {
        throw new Error("Invalid OTP");
      }

      // OTP is valid, clean up and generate token
      await supabase
        .schema("ayul")
        .from("otp_verifications")
        .delete()
        .eq("phone", normalizedPhone);
      const token = generateToken(normalizedPhone);

      return new Response(
        JSON.stringify({ success: true, message: "OTP verified", token }),
        {
          status: 200,
          headers: { "Content-Type": "application/json", ...CORS_HEADERS },
        }
      );
    } else {
      throw new Error("Invalid or unsupported action");
    }
  } catch (error) {
    console.error("Server error:", error);

    return new Response(
      JSON.stringify({
        error: {
          http_code: 500,
          message: error.message || "Internal Server Error",
          details: error.stack || "No additional details",
        },
      }),
      {
        status: 500,
        headers: { "Content-Type": "application/json", ...CORS_HEADERS },
      }
    );
  }
});
this code I am using I face the issue session is not active in login
